Add bazel wrapper script

keith opened the pull request:
This script works for macOS and Linux

keith committed:
Add bazel wrapper script

This script works for macOS and Linux

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
`#!/usr/bin/env bash`?

keith replied:
Same reason as before:

```
% /usr/bin/env bash --version
GNU bash, version 4.4.23(1)-release (x86_64-apple-darwin18.0.0)
% /bin/bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin18)
```

We want to have a more stable version of bash for everyone

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+
+set -euo pipefail
+
+readonly bazel_version="0.19.0"
```
I'd recommend reading this from another file, so that users of the wrapper script won't need to modify it just to update the Bazel version

For reference, this is how Gradle and Buck (OkBuck) wrapper work

keith replied:
What's the problem with modifying this when the version changes? Since it isn't generated. If we did that we'd also have to do that with the sha's and the platform conditional

artem-zinnatullin replied:
Allows easier upgrades of the script, as discussed locally it's a reference rather than an implementation, so I guess it's fine

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+
+readonly bazel_version="0.19.0"
+if [[ $OSTYPE == darwin* ]]; then
+  readonly bazel_version_sha="2a3e309fe359bd8beca18b2ab095969208df1c73bc4002d92341d0fff83b4a01"
```
nit: rename to something like `bazel_binary_shasum`?

I thought that was a commit sha until I figured that it has to be check sum

keith replied:
done

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+  readonly bazel_version_url="https://github.com/bazelbuild/bazel/releases/download/${bazel_version}/bazel-${bazel_version}-installer-darwin-x86_64.sh"
+else
+  readonly bazel_version_sha="b9e988643b10444d6d8eb04461418b7b80e3d171e12d1c8b6e2a50f541e739e7"
+  readonly bazel_version_url="https://github.com/bazelbuild/bazel/releases/download/${bazel_version}/bazel-${bazel_version}-installer-linux-x86_64.sh"
```
The url can also be extracted into a file, but that's a nit (use-case is when you have a fork)

keith replied:
What's the benefit?

artem-zinnatullin replied:
Similar to having version detached from the script itself, data vs code (discussed above, it's a nit)

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+fi
+
+readonly script_root="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+readonly bazel_root="$script_root/tmp/bazel/versions/$bazel_version"
```
Maybe go with `$script_root/.bazel` or something like this? 

`tmp` might be confusing given that files will stay there for quite some time and also it doesn't start with `.` thus will be visible in file manager by default

For reference, that's what Gradle and Buck (OkBuck) wrapper do

artem-zinnatullin replied:
Correction: Gradle wrapper stores Gradle binaries in `$HOME/.gradle`, which allows to reuse them between different projects without redownloading

keith replied:
I'm intentionally storing this in the same repo. I could do `.bazel` for this, in our repo we use `tmp` for many things so it made sense to put here

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+readonly bazel_executable="$bazel_root/bin/bazel"
+
+if [[ ! -x "$bazel_executable" ]]; then
+  echo "Installing bazel..." >&2
```
Downloading?

artem-zinnatullin replied:
Also, `Bazel` with "B" lol

keith replied:
I don't log again between downloading and installing, but I could. I like that the log is the same as the command line binary name

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+
+  curl \
+    --show-error \
+    --silent \
```
You might actually want to show progress, curl is pretty ok at doing that

usecase: I'm on mobile network where each MB costs $ and I might want to realize that I'm downloading ~150Mb of stuff sooner than later

keith replied:
yea, we do for ours, we only make it silent on CI, but for this script i didn't want to have an environment variable that's specific to us for that. alternatively I could remove this and people could deal with the other problem

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+    --output "$script_path" \
+    "$bazel_version_url"
+
+  if echo "$bazel_version_sha  $script_path" | shasum --check --status; then
```
why double space between version sha and script path?

keith replied:
that's the format shasum expects

artem-zinnatullin replied:
oh lol

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,39 @@
+    --output "$script_path" \
+    "$bazel_version_url"
+
+  if echo "$bazel_version_sha  $script_path" | shasum --check --status; then
```
btw, shasum check happens every time you run `bazelw` even if it was already downloaded, is this intended?

shasum for Bazel binary seems to take around 0.4s on i9 mbp, a bit costly

artem-zinnatullin replied:
nvm, misread it, it happens only at install time

keith committed:
Rename with shasum

keith committed:
Make curl show progress

This likely needs to be disabled on user's CI systems, but that's up to
their specific configuration to do

artem-zinnatullin commented on tools/bazelw:
```diff
@@ -0,0 +1,38 @@
+readonly bazel_version="0.19.0"
+if [[ $OSTYPE == darwin* ]]; then
+  readonly bazel_version_shasum="2a3e309fe359bd8beca18b2ab095969208df1c73bc4002d92341d0fff83b4a01"
+  readonly bazel_version_url="https://github.com/bazelbuild/bazel/releases/download/${bazel_version}/bazel-${bazel_version}-installer-darwin-x86_64.sh"
```
nit: looks like non-sh binary works too

keith commented:
Almost this example is checked in here now https://github.com/lyft/envoy-mobile/blob/master/bazelw
